### PR TITLE
[CSL-1341] Move tx checks

### PR DESCRIPTION
**Pull request by pva701**

### CODE REVIEW

**Comment by pva701** [comment]
@gromakovsky you can review, I am going to add `GenesisStakeholders`

**Comment by pva701** [comment]
this PR should be merged after 
https://github.com/input-output-hk/cardano-sl/pull/1184
because it contains ^

**Comment by pva701** [comment]
@gromakovsky can you review?

**Comment by gromakovsky** [comment]
I am going to review #1184 first.

File: txp/Pos/Txp/Toil/Types.hs [hunk 1]
```diff
@@ -55,6 +61,13 @@ import qualified Pos.Util.Modifier          as MM
 -- output) pairs.
 type Utxo = Map TxIn TxOutAux
 
+-- | Convert 'Utxo' to 'StakesMap'.
+utxoToStakes :: Utxo -> StakesMap
```
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Types.hs]: Why is it in `Types`?

File: txp/Pos/Txp/Toil/Types.hs [hunk 2]
```diff
@@ -64,7 +77,21 @@ utxoF :: Format r (Utxo -> r)
 utxoF = later formatUtxo
 
 -- | Wrapper for genesis utxo.
-newtype GenesisUtxo = GenesisUtxo { unGenesisUtxo :: Utxo }
+newtype GenesisUtxo = GenesisUtxo
+    { unGenesisUtxo :: Utxo
+    }
+
+data GenesisTxpContext = UnsafeGenesisTxpContext
```
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Types.hs]: Add a comment.

File: txp/Pos/Txp/Toil/Types.hs (hunk 1 above)
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Types.hs]: Hm, because it's needed for `mkGenesisTxpContext`, okay.

File: txp/Pos/Txp/Toil/Types.hs [hunk 3]
```diff
@@ -64,7 +77,21 @@ utxoF :: Format r (Utxo -> r)
...
+    { _gtcUtxo         :: !GenesisUtxo
+    , _gtcStakeholders :: !GenesisStakeholders
+    }
+
+makeLenses ''GenesisTxpContext
+
+mkGenesisTxpContext :: Utxo -> GenesisTxpContext
+mkGenesisTxpContext utxo = UnsafeGenesisTxpContext
+    (GenesisUtxo utxo)
+    (GenesisStakeholders . getKeys . utxoToStakes $ utxo)
```
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Types.hs]: Actually it's inefficient, but I don't mind because it should be done only on start.

File: txp/Pos/Txp/Toil/Logic.hs [hunk 4]
```diff
@@ -81,116 +103,154 @@ rollbackToil txun = do
...
-        -- 1. Reject all new-coming transactions (b/c we can't calculate
-        --    fee for them)
-        -- 2. Use latest policy of known type
-        -- 3. Discard the check
-        -- Implementation-wise, the 1st option corresponds to throwing an
-        -- error here (reject), the 3rd option -- doing nothing (accept), and
-        -- the 2nd option would require some engineering feats to
-        -- retrieve previous 'TxFeePolicy' and check against it.
-        return ()
+verifyTxFeePolicy (TxFee txFee) policy txSize = do
```
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Logic.hs]: Absolutely useless change.

File: txp/Pos/Txp/Toil/Logic.hs [hunk 5]
```diff
@@ -17,33 +21,51 @@ module Pos.Txp.Toil.Logic
...
 
 ----------------------------------------------------------------------------
 -- Global
 ----------------------------------------------------------------------------
 
-type GlobalToilMode m = ( MonadUtxo m
-                        , MonadBalances m
-                        , MonadToilEnv m
-                        , WithLogger m)
+type GlobalApplyToilMode ctx m =
```
Comment (gromakovsky) [inline txp/Pos/Txp/Toil/Logic.hs]: `ctx` is useless here.

File: txp/Pos/Txp/Logic/Local.hs [hunk 6]
```diff
@@ -79,68 +74,48 @@ txProcessTransaction itw@(txId, txAux) = do
...
-       , MonadTxpMem () ctx m
-       , WithLogger m
-       )
+txNormalize
+    :: ( TxpLocalWorkMode ctx m
+       , MonadSlots m)
     => m ()
 txNormalize = do
     utxoTip <- GS.getTip
+    epoch <- maybe (throwM ToilUnknownCurEpoch) (pure . siEpoch) =<< getCurrentSlot
```
Comment (gromakovsky) [inline txp/Pos/Txp/Logic/Local.hs]: No, it's broken. When we load blocks after long offline, we don't know the current slot. This function will throw and something will go wrong.
The proper solution is to just clean mempool if we don't know the current slot.

File: txp/Pos/Txp/Logic/Global.hs [hunk 7]
```diff
@@ -45,13 +48,14 @@ txpGlobalSettings =
...
 verifyBlocks
-    :: forall m.
-       TxpGlobalVerifyMode m
+    :: forall ctx m.
+       TxpGlobalVerifyMode ctx m
     => Bool -> OldestFirst NE TxpBlock -> m (OldestFirst NE TxpUndo)
-verifyBlocks verifyAllIsKnown newChain =
-    fst <$> runToilAction @_ @() (mapM verifyDo newChain)
+verifyBlocks verifyAllIsKnown newChain = do
+    epoch <- siEpoch <$> (note ToilUnknownCurEpoch =<< getCurrentSlot)
```
Comment (gromakovsky) [inline txp/Pos/Txp/Logic/Global.hs]: What is it? It's completely broken. You should take epoch from header.

**Comment by gromakovsky** [comment]
#1184 is merged. Rebase please.

File: txp/Pos/Txp/Logic/Global.hs (hunk 7 above)
Comment (pva701) [inline txp/Pos/Txp/Logic/Global.hs]: ok. don't we want to reflect in the name of function the fact that only blocks from same epoch can be passed? or this invariant is true for every block processing function?

**Comment by pva701** [comment]
@gromakovsky rebased

File: txp/Pos/Txp/Logic/Global.hs (hunk 7 above)
Comment (gromakovsky) [inline txp/Pos/Txp/Logic/Global.hs]: > ok. don't we want to reflect in the name of function the fact that only blocks from same epoch can be passed?

Well, I think it would be useful, but it's minor.

> this invariant is true for every block processing function?

Tt is true for every function except all rollback functions and `verifyAndApplyBlocks`.

File: txp/Pos/Txp/Settings/Global.hs [hunk 8]
```diff
@@ -37,7 +37,6 @@ type TxpCommonMode ctx m =
 type TxpGlobalVerifyMode ctx m =
     ( TxpCommonMode ctx m
     , MonadError ToilVerFailure m
-    , MonadSlots m  -- TODO: I don't like it (@gromak)
```
Comment (gromakovsky) [inline txp/Pos/Txp/Settings/Global.hs]: Did you test whether `WITH_EXPLORER` build still works after this change? I suspect it doesn't.

File: db/Pos/DB/Class.hs [hunk 9]
```diff
@@ -183,13 +187,14 @@ gsMaxProposalSize = bvdMaxProposalSize <$> gsAdoptedBVData
 gsUnlockStakeEpoch :: MonadGState m => m EpochIndex
 gsUnlockStakeEpoch = bvdUnlockStakeEpoch <$> gsAdoptedBVData
 
+-- | Checks if provided epoch is in the bootstrap era (pure version)
+isBootstrapEraPure :: EpochIndex -> EpochIndex -> Bool
+isBootstrapEraPure unlockEpoch curEpoch =
+    not isDevelopment && isBootstrapEra unlockEpoch curEpoch
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: It's absolutely bullshit, because there is `isBootstrapEra` (and `not isDevelopment` should be ~removed~ there), but I am going to do CSL-1292 today, so I will fix it later.

File: db/Pos/DB/Class.hs [hunk 10]
```diff
@@ -183,13 +187,14 @@ gsMaxProposalSize = bvdMaxProposalSize <$> gsAdoptedBVData
...
+
 -- | Checks if provided epoch is in the bootstrap era.
 gsIsBootstrapEra :: MonadGState m => EpochIndex -> m Bool
-gsIsBootstrapEra epoch =
-    if isDevelopment then pure False
-    else do
-        unlockStakeEpoch <- gsUnlockStakeEpoch
-        return $ isBootstrapEra unlockStakeEpoch epoch
+gsIsBootstrapEra epoch = flip isBootstrapEraPure epoch <$> gsUnlockStakeEpoch
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: I like `do` version slightly more, because it makes it harder to confuse the order of arguments. But it's minor.

File: txp/Pos/Txp/Settings/Global.hs (hunk 8 above)
Comment (pva701) [inline txp/Pos/Txp/Settings/Global.hs]: Lol, in the master there isn't this constraint and it works (I suppose).
I just returned this constraint to `master` state, why wouldn't it work?

File: src/Pos/Client/Txp/History.hs [hunk 11]
```diff
@@ -219,6 +219,7 @@ type TxHistoryEnv ctx m =
     , MonadSlots m
     , MonadReader ctx m
     , HasLens GenesisUtxo ctx GenesisUtxo
```
Comment (gromakovsky) [inline src/Pos/Client/Txp/History.hs]: Is this lens still needed, btw? Can you check?

File: txp/Pos/Txp/Settings/Global.hs (hunk 8 above)
Comment (pva701) [inline txp/Pos/Txp/Settings/Global.hs]: ok, I got it. will try now

File: src/Pos/Block/Logic/Internal.hs [hunk 12]
```diff
@@ -10,8 +10,10 @@ module Pos.Block.Logic.Internal
          -- * Constraints
          MonadBlockBase
        , MonadBlockVerify
+       , MonadBlockApplyDB
        , MonadBlockApply
 
+       , applyBlocksDbUnsafe
```
Comment (gromakovsky) [inline src/Pos/Block/Logic/Internal.hs]: I totally don't like new names in this module. They are absolutely confusing. I would export `applyBlocksUnsafe` which doesn't do normalization and would define `normalizeBlocks` in `VAR` and use it in `verifyAndApplyBlocks` in the end. It should also resolve CSL-544.

File: src/Pos/Context/Functions.hs [hunk 13]
```diff
@@ -4,6 +4,7 @@ module Pos.Context.Functions
        (
          -- * Genesis
          GenesisUtxo(..)
+       , mkGenesisTxpContext
```
Comment (gromakovsky) [inline src/Pos/Context/Functions.hs]: I don't think it should be here.

File: src/Pos/Generator/Block/Mode.hs [hunk 14]
```diff
@@ -123,7 +124,7 @@ data BlockGenContext = BlockGenContext
     , bgcSystemStart       :: !Timestamp
     , bgcParams            :: !BlockGenParams
     , bgcDelegation        :: !DelegationVar
-    , bgcGenesisUtxo       :: !GenesisUtxo
+    , bgcGenesisTxpContext :: !GenesisTxpContext
```
Comment (gromakovsky) [inline src/Pos/Generator/Block/Mode.hs]: I think it's wrong, it should have only `GenesisStakeholders`. I don't see why `GenesisUtxo` is needed. Moreover, it seems dangerous to have it in the context, because someone can accidentally use it, but it's wrong.
Maybe I am missing smth tho.

File: src/Pos/Generator/Block/Mode.hs [hunk 15]
```diff
@@ -182,7 +183,7 @@ mkBlockGenContext bgcParams@BlockGenParams{..} = do
         return BlockGenContext {..}
   where
     -- Genesis utxo is needed only for boot era stakeholders
-    bgcGenesisUtxo =
+    bgcGenesisTxpContext =
```
Comment (gromakovsky) [inline src/Pos/Generator/Block/Mode.hs]: We both know that it's bullshit, can you remove it? `GenesisStakeholders` should come from `MonadBlockGen` instead.

File: src/Pos/WorkMode/Class.hs [hunk 16]
```diff
@@ -93,6 +94,7 @@ type WorkMode ssc ctx m
       , HasLens SecurityParams ctx SecurityParams
       , HasLens TxpGlobalSettings ctx TxpGlobalSettings
       , HasLens GenesisUtxo ctx GenesisUtxo
+      , HasLens GenesisStakeholders ctx GenesisStakeholders
```
Comment (gromakovsky) [inline src/Pos/WorkMode/Class.hs]: I think it's better to have `HasLens GenesisTxpContext` instead, but it's minor.

Comment (gromakovsky) [inline src/Pos/WorkMode/Class.hs]: (instead of these two)

File: src/Pos/Launcher/Param.hs [hunk 17]
```diff
@@ -61,7 +62,7 @@ data NodeParams = NodeParams
     , npSecretKey      :: !SecretKey            -- ^ Primary secret key of node
     , npUserSecret     :: !UserSecret           -- ^ All node secret keys
     , npBaseParams     :: !BaseParams           -- ^ See 'BaseParams'
-    , npGenesisUtxo    :: !GenesisUtxo          -- ^ Predefined genesis utxo
+    , npGenesisTxpCont :: !GenesisTxpContext    -- ^ Predefined genesis context related to txp data.
```
Comment (gromakovsky) [inline src/Pos/Launcher/Param.hs]: We use `ctx` as a short version of `context`, don't use `cont`, it's not conventional.

File: test/Test/Pos/Block/Logic/Mode.hs [hunk 18]
```diff
@@ -136,7 +137,7 @@ instance Buildable TestParams where
             _tpStakeDistribution
             _tpStartTime
       where
-        utxo = _tpGenUtxo & \(GenesisUtxo u) -> u
+        utxo = _tpGenTxpContext & \g -> unGenesisUtxo (g ^. gtcUtxo)
```
Comment (gromakovsky) [inline test/Test/Pos/Block/Logic/Mode.hs]: Without pattern matching such construct doesn't make any sense. You can write `unGenesisUtxo (_tpGenTxpContext ^. gtcUtxo)`.

File: src/Pos/Context/Functions.hs (hunk 13 above)
Comment (pva701) [inline src/Pos/Context/Functions.hs]: otherwise, I would be forced to add `cardano-sl-txp` to block-gen or something like that.
this functions more or less related to `Context`

Comment (gromakovsky) [inline src/Pos/Context/Functions.hs]: > I would be forced to add cardano-sl-txp to block-gen or something like that.

You are saying it like it's something bad.

File: src/Pos/Generator/Block/Mode.hs (hunk 14 above)
Comment (pva701) [inline src/Pos/Generator/Block/Mode.hs]: I am convinced that there are places where `GenesisUtxo` is needed, for example, `Tracking.hs`

Comment (gromakovsky) [inline src/Pos/Generator/Block/Mode.hs]: It doesn't seem right to use `Tracking.hs` from block generator.

Comment (pva701) [inline src/Pos/Generator/Block/Mode.hs]: didn't notice that it's block-gen, ok, I can remove. but it can be needed in the future

File: src/Pos/Context/Functions.hs (hunk 13 above)
Comment (pva701) [inline src/Pos/Context/Functions.hs]: @gromakovsky can you explain, why `Context` can re-export `GenesisUtxo`, but `Functions` can't re-export `mkGenesisTxpContext`?

Comment (gromakovsky) [inline src/Pos/Context/Functions.hs]: I think `Context` re-exports `GenesisUtxo` because it was initially defined there and then someone didn't want to change imports in many places.
Anyway, that's not the main reason I rejected this PR, if it was the only one I would approve because it's not so important.

File: src/Pos/Block/Logic/Internal.hs (hunk 12 above)
Comment (pva701) [inline src/Pos/Block/Logic/Internal.hs]: what should functions call which called `verifyBlocksUnsafe` earlier?

File: src/Pos/Generator/Block/Mode.hs (hunk 14 above)
Comment (gromakovsky) [inline src/Pos/Generator/Block/Mode.hs]: It's better to remove. If later someone thinks they need it, they will have to think. And maybe it will turn out that it's not needed.

File: src/Pos/Block/Logic/Internal.hs (hunk 12 above)
Comment (gromakovsky) [inline src/Pos/Block/Logic/Internal.hs]: What is `verifyBlocksUnsafe`?

Comment (gromakovsky) [inline src/Pos/Block/Logic/Internal.hs]: My point is that we should separate `applyBlocksUnsafe` and `normalize` and let the caller decide whether they need to call both or only one. Because actually there is only one place where `normalize` should be done, in other places it's only needed to apply.

File: core/Pos/Util/Util.hs [hunk 19]
```diff
@@ -325,7 +325,7 @@ type HasLens' s a = HasLens a s a
 
 -- | Version of 'lensOf' which is used when lens is to the same type
 -- as the tag.
-lensOf' :: forall s a. HasLens' s a => Lens' s a
+lensOf' :: forall a s. HasLens' s a => Lens' s a
```
Comment (volhovM) [inline core/Pos/Util/Util.hs]: Such a change.

File: src/Pos/Block/Logic/Internal.hs [hunk 20]
```diff
@@ -79,17 +82,29 @@ type MonadBlockBase ssc ctx m
        -- Needed by some components.
        , MonadGState m
        -- This constraints define block components' global logic.
-       , HasLens TxpGlobalSettings ctx TxpGlobalSettings
        , HasLens LrcContext ctx LrcContext
+       , HasLens TxpGlobalSettings ctx TxpGlobalSettings
        , SscGStateClass ssc
+       , HasLens GenesisStakeholders ctx GenesisStakeholders
+       , MonadDelegation ctx m -- TODO: split dlgApplyBlocks blocks
```
Comment (volhovM) [inline src/Pos/Block/Logic/Internal.hs]: Wat?

File: src/Pos/Block/Logic/Internal.hs [hunk 21]
```diff
@@ -79,17 +82,29 @@ type MonadBlockBase ssc ctx m
...
 type MonadBlockApply ssc ctx m
+     = ( MonadBlockApplyDB ssc ctx m
+       -- Needed for normalization.
+       , MonadTxpMem TxpExtra_TMP ctx m
+       , SscLocalDataClass ssc
+       , HasLens UpdateContext ctx UpdateContext
+       )
+
+-- | Set of constraints necessary to apply or rollback blocks at high-level (without normalization).
+type MonadBlockApplyDB ssc ctx m
```
Comment (volhovM) [inline src/Pos/Block/Logic/Internal.hs]: Reverse definition order is posos. I'd swap `MonadBlockApplyDB` and `MonadBlockApply`. And rename, yes.

File: src/Pos/Context/Functions.hs (hunk 13 above)
Comment (volhovM) [inline src/Pos/Context/Functions.hs]: Looks weird indeed. `Context` reexports `GenesisUtxo` because it was the part of the context. And it is. I guess.

File: src/Pos/Explorer/Txp/Local.hs [hunk 22]
```diff
@@ -130,9 +136,9 @@ eTxProcessTransaction itw@(txId, TxAux {taTx = UnsafeTx {..}}) = do
                      flip evalUtxoStateT resolved $
                      flip runReaderT eet $
                      runExplorerReaderWrapper $
+                     flip runReaderT genStks $
```
Comment (volhovM) [inline src/Pos/Explorer/Txp/Local.hs]: Eeh, i'd prefer a specific reader environment with one datatype instead of two `ReaderT` layers.

File: txp/Pos/Txp/Logic/Local.hs [hunk 23]
```diff
@@ -77,68 +75,49 @@ txProcessTransaction itw@(txId, txAux) = do
...
+       , MonadSlots m)
     => m ()
-txNormalize = do
-    utxoTip <- GS.getTip
-    localTxs <- getLocalTxs
-    ToilModifier {..} <-
-        runDBToil $ execToilTLocal mempty def mempty $ normalizeToil localTxs
-    setTxpLocalData "txNormalize" (_tmUtxo, _tmMemPool, _tmUndos, utxoTip, _tmExtra)
+txNormalize = getCurrentSlot >>= \case
+    Nothing -> clearTxpMemPool "txNormalize "
```
Comment (volhovM) [inline txp/Pos/Txp/Logic/Local.hs]: What does this mean?

File: txp/Pos/Txp/Toil/Failure.hs [hunk 24]
```diff
@@ -42,6 +42,7 @@ data ToilVerFailure
                           , tifSize   :: !Byte }
     | ToilUnknownAttributes !ByteString
     | ToilBootDifferentStake !TxOutDistribution
+    | ToilUnknownCurEpoch
```
Comment (volhovM) [inline txp/Pos/Txp/Toil/Failure.hs]: Isn't there `ToilSlotUnknown` for this already?

File: src/Pos/Block/Logic/Internal.hs (hunk 21 above)
Comment (gromakovsky) [inline src/Pos/Block/Logic/Internal.hs]: I'd rename it back to `MonadBlockApply` and delete current `MonadBlockApply`.

File: src/Pos/Explorer/Txp/Local.hs (hunk 22 above)
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Local.hs]: Also nobody knows if this code compiles :)

File: txp/Pos/Txp/Logic/Local.hs (hunk 23 above)
Comment (gromakovsky) [inline txp/Pos/Txp/Logic/Local.hs]: It means that we delete all transactions from the mempool if we don't know the current slot. Which is fine because it means we are lost. What I don't like is that we don't update tip. It's bad @pva701

File: src/Pos/Context/Functions.hs (hunk 13 above)
Comment (pva701) [inline src/Pos/Context/Functions.hs]: >Context reexports GenesisUtxo because it was the part of the context
sound like the most strange thing I have ever heard, lets reexprot all datatypes from `Context.hs` then. gromak's motivation sounds more or less reasonable.

`mkGenesisTxpContext` creates the part of context, why not reexprot?

File: core/Pos/Util/Util.hs (hunk 19 above)
Comment (pva701) [inline core/Pos/Util/Util.hs]: lol, it's not strange.
it's for convenient to avoid such code `lensOf' @_ @GenesisUtxo` and write instead `lensOf' @GenesisUtxo`

File: src/Pos/Block/Logic/Internal.hs (hunk 20 above)
Comment (pva701) [inline src/Pos/Block/Logic/Internal.hs]: @volhovM delegation applies blocks and normalises in memory state at the same time.
I propose to split it

File: src/Pos/Explorer/Txp/Local.hs (hunk 22 above)
Comment (pva701) [inline src/Pos/Explorer/Txp/Local.hs]: @volhovM I tried to do it, but there is not `HasLens` for first and second elements of pair

File: txp/Pos/Txp/Toil/Failure.hs (hunk 24 above)
Comment (pva701) [inline txp/Pos/Txp/Toil/Failure.hs]: make sense

File: src/Pos/Client/Txp/History.hs (hunk 11 above)
Comment (pva701) [inline src/Pos/Client/Txp/History.hs]: still needed

File: src/Pos/Block/Logic/Internal.hs (hunk 20 above)
Comment (volhovM) [inline src/Pos/Block/Logic/Internal.hs]: For what reason?

Comment (pva701) [inline src/Pos/Block/Logic/Internal.hs]: because we introduce `normalize` (normalisation of mempool as separate function), it's not part of `applyBlocksUnsafe` anymore

Comment (pva701) [inline src/Pos/Block/Logic/Internal.hs]: but I see the comment, that it's not so easy to normalise dlg, I will remove my comment

Comment (volhovM) [inline src/Pos/Block/Logic/Internal.hs]: My point was:
1. Normalization while doing application goes super natural for dlg
2. Maybe there are reasons to separate them though i don't see them (might have missed the discussion).

File: src/Pos/Explorer/Txp/Local.hs (hunk 22 above)
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Local.hs]: 🤦‍♂️ 
The proper way is to create a special type and define `Has` for it.

Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Local.hs]: This way is ugly though 😞

Comment (pva701) [inline src/Pos/Explorer/Txp/Local.hs]: >The proper way is to create a special type and define Has for it.
делать мне больше нечего, чтобы на каждый чих, ради того что кому-то там два ридера не нравится создавать дататайп :peka:

File: src/Pos/Block/Logic/Internal.hs (hunk 20 above)
Comment (gromakovsky) [inline src/Pos/Block/Logic/Internal.hs]: One reason is that we need to do normalization only once when we apply a big sequence of blocks, but currently we can do it multiple times. It's not very important for delegation I think.

**Comment by pva701** [comment]
@gromakovsky review plz

File: src/Pos/Generator/Block/Mode.hs (hunk 15 above)
Comment (gromakovsky) [inline src/Pos/Generator/Block/Mode.hs]: Remove bullshit please 😔

**Comment by pva701** [comment]
@gromakovsky let's remove bullshit in next issues

**Comment by pva701** [comment]
@gromakovsky can you re-review again?

File: db/Pos/DB/Class.hs [hunk 25]
```diff
@@ -187,7 +190,7 @@ gsUnlockStakeEpoch = bvdUnlockStakeEpoch <$> gsAdoptedBVData
 gsIsBootstrapEra :: MonadGState m => EpochIndex -> m Bool
 gsIsBootstrapEra epoch = do
     unlockStakeEpoch <- gsUnlockStakeEpoch
-    return $ isBootstrapEra unlockStakeEpoch epoch
+    pure $ isBootstrapEra unlockStakeEpoch epoch
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: Such a change.

**Comment by gromakovsky** [comment]
Did you test compilation with explorer?

File: db/Pos/DB/Class.hs (hunk 25 above)
Comment (pva701) [inline db/Pos/DB/Class.hs]: pure <3

**Comment by pva701** [comment]
I did, it's ok

File: src/Pos/Explorer/Txp/Local.hs [hunk 26]
```diff
@@ -147,22 +153,19 @@ eTxProcessTransaction itw@(txId, TxAux {taTx = UnsafeTx {..}}) = do
...
-       , MonadGState m
-       , MonadTxpMem ExplorerExtra ctx m
-       , WithLogger m
+       ( ETxpLocalWorkMode ctx m
+       , MonadSlots m
        )
     => m ()
 eTxNormalize = do
     utxoTip <- GS.getTip
+    epoch <- maybe (throwM ToilSlotUnknown) (pure . siEpoch) =<< getCurrentSlot
```
Comment (gromakovsky) [inline src/Pos/Explorer/Txp/Local.hs]: Wait, what is it? Why is it different from `txNormalize`? It should be the same.

File: src/node/Params.hs [hunk 27]
```diff
@@ -94,9 +96,11 @@ getNodeParams args@Args {..} systemStart = do
...
                 (CLI.richPoorDistr commonArgs)
                 (CLI.expDistr commonArgs)
-    let npGenesisUtxo
-            | isDevelopment = genesisUtxo Nothing (devAddrDistr devStakeDistr)
-            | otherwise = genesisUtxoProduction
+    let npGenesisTxpCtx = mkGenesisTxpContext $
+            if isDevelopment
+            then genesisUtxo Nothing (devAddrDistr devStakeDistr)
+            else genesisUtxo (Just genesisProdBootStakeholders)
+                                genesisProdAddrDistribution
```
Comment (gromakovsky) [inline src/node/Params.hs]: Why did you change `genesisUtxoProduction` to `genesisUtxo (Just genesisProdBootStakeholders) genesisProdAddrDistribution`?

File: src/Pos/Explorer/Txp/Local.hs (hunk 26 above)
Comment (pva701) [inline src/Pos/Explorer/Txp/Local.hs]: FUCK this FUCKING boilerplate, It's so tedious to manage same code in the fucking two places.

**Comment by domenkozar** [comment]
@pva701 can you rebase on top of master?
